### PR TITLE
Parse tree visitor

### LINKED ISSUE #18: Implement a parse tree visitor
The parser branch already has a parse tree, but we will need a visitor base class to be able to write recursion schemes simpler

### PR DESCRIPTION
**Pull request by thinker227**
Closes #18. Implement a visitor interface and base visitor implementation for `ParseTree`.

### CODE REVIEW

File: src/Draco.Compiler/Syntax/ParseTreeVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    T VisitUnaryExpr(Expr.Unary expr);
+    T VisitBinaryExpr(Expr.Binary expr);
+    T VisitGroupingExpr(Expr.Grouping expr);
+}
+
+/// <summary>
+/// Provides a base implementation of <see cref="IParseTreeVisitor{T}"/>.
+/// </summary>
+/// <typeparam name="T">The return type of the visitor.</typeparam>
+internal abstract class BaseParseTreeVisitor<T> : IParseTreeVisitor<T>
```
Comment (LPeter1997) [inline src/Draco.Compiler/Syntax/ParseTreeVisitor.cs:54]: Isn't a Base suffix the convention? So `ParseTreeVisitorBase`? I'm honestly not sure

File: src/Draco.Compiler/Syntax/ParseTreeVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    public abstract T VisitGotoExpr(Expr.Goto expr);
+    public abstract T VisitReturnExpr(Expr.Return expr);
+    public abstract T VisitLiteralExpr(Expr.Literal expr);
+    public abstract T VisitFuncCallExpr(Expr.FuncCall expr);
+    public abstract T VisitIndexExpr(Expr.Index expr);
+    public abstract T VisitVariableExpr(Expr.Variable expr);
+    public abstract T VisitMemberAccessExpr(Expr.MemberAccess expr);
+    public abstract T VisitUnaryExpr(Expr.Unary expr);
+    public abstract T VisitBinaryExpr(Expr.Binary expr);
+    public abstract T VisitGroupingExpr(Expr.Grouping expr);
```
Comment (LPeter1997) [inline src/Draco.Compiler/Syntax/ParseTreeVisitor.cs:146]: Why do these don't do their recursive thing?

File: src/Draco.Compiler/Syntax/ParseTreeVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    public virtual T VisitStmt(Stmt stmt) => stmt switch
+    {
+        Stmt.Decl decl => this.VisitDeclStmt(decl),
+        Stmt.Expr expr => this.VisitExprStmt(expr),
+
+        _ => throw new InvalidOperationException()
+    };
+    public abstract T VisitDeclStmt(Stmt.Decl stmt);
+    public abstract T VisitExprStmt(Stmt.Expr stmt);
```
Comment (LPeter1997) [inline src/Draco.Compiler/Syntax/ParseTreeVisitor.cs:114]: Same, as well as for the other abstract methods, shouldn't they recursively visit their subelements?

File: src/Draco.Compiler/Syntax/ParseTreeVisitor.cs (hunk 2 above)
Comment (thinker227) [reply src/Draco.Compiler/Syntax/ParseTreeVisitor.cs:146]: Sure these could visit all of their subexpressions, but what value should they return? `default(T)`?

Comment (thinker227) [reply src/Draco.Compiler/Syntax/ParseTreeVisitor.cs:146]: Also I'm assuming you'd have to call `base.VisitWhatever(whatever)` if you want to override any of these?

Comment (LPeter1997) [reply src/Draco.Compiler/Syntax/ParseTreeVisitor.cs]: `default(T)` or a virtual default providing prop are both fine. And yes, you need to call base to recurse, just like ANTLR visitors

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    - name: Setup .NET
+      uses: actions/setup-dotnet@v1
+      with:
+        dotnet-version: ${{ env.DOTNET_VERSION }}
+
+    - name: Install dependencies
+      run: dotnet restore ./src
+
+    - name: Build
+      run: dotnet build ./src --configuration ${{ env.CONFIGURATION }} --no-restore
```
Comment (LPeter1997) [inline .github/workflows/build.yaml:41]: How did these get in again? 😅 
Try to pull the parser branch from this origin.

File: src/Draco.Compiler/Syntax/ParseTreeVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    T VisitGroupingExpr(Expr.Grouping expr);
+}
+
+/// <summary>
+/// Provides a base implementation of <see cref="IParseTreeVisitor{T}"/>.
+/// </summary>
+/// <typeparam name="T">The return type of the visitor.</typeparam>
+internal abstract class BaseParseTreeVisitor<T> : IParseTreeVisitor<T>
+{
+    public virtual T VisitNode(ParseTree node) => node switch
```
Comment (LPeter1997) [inline src/Draco.Compiler/Syntax/ParseTreeVisitor.cs:56]: Blank lines between close braces and signatures are a little inconsistent, I'd suggest just having the newline everywhere.

File: src/Draco.Compiler/Syntax/ParseTreeVisitor.cs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+        _ => throw new InvalidOperationException()
+    };
+
+    public virtual T VisitCompilationUnit(CompilationUnit compilationUnit)
+    {
+        foreach (var decl in compilationUnit.Declarations)
+        {
+            this.VisitDecl(decl);
+        }
```
Comment (LPeter1997) [inline src/Draco.Compiler/Syntax/ParseTreeVisitor.cs:75]: You could have helpers for visiting sequences.

File: src/Draco.Compiler/Syntax/ParseTreeVisitor.cs [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    {
+        foreach (var param in decl.Params.Value.Elements)
+        {
+            this.VisitFuncParam(param.Value);
+        }
+
+        if (decl.Type is not null)
+        {
+            this.VisitTypeSpecifier(decl.Type);
+        }
```
Comment (LPeter1997) [inline src/Draco.Compiler/Syntax/ParseTreeVisitor.cs:98]: Same for nullables, a helper might not be all evil

File: .github/workflows/build.yaml (hunk 4 above)
Comment (thinker227) [reply .github/workflows/build.yaml:41]: Very, very good question. Git just did not want to cooperate.

**Pull request merged**